File: ch5/clip7/boj15486/main.cpp
#include <iostream>

int main()
{
    int N, T, P;
    int dp[1500002];

    scanf("%d", &N);

    for(int i = 0; i <= N + 1; i++)
    {
        dp[i] = 0;
    }

    for(int i = 1; i <= N; i++)
    {
        scanf("%d %d", &T, &P);

        if(dp[i] < dp[i - 1])
        {
            dp[i] = dp[i - 1];
        }

        if(i + T <= N + 1)
        {
            if(dp[i + T] < dp[i] + P)
            {
                dp[i + T] = dp[i] + P;
            }
        }
    }

    printf("%d", std::max(dp[N], dp[N + 1]));

    return 0;
}

File: ch5/clip7/boj15486/Main.java
import java.util.Scanner;

public class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int t[] = new int[n + 2];
        int p[] = new int[n + 2];
        int d[] = new int[n + 2];
        for(int i = 1; i <= n; i++) {
            t[i] = sc.nextInt();
            p[i] = sc.nextInt();
        }
        for(int i = 1; i <= n + 1; i++) {
            d[i] = Math.max(d[i-1], d[i]);

            if(i + t[i] > n + 1) continue;
            d[i + t[i]] = Math.max(d[i + t[i]], d[i] + p[i]);
        }
        System.out.println(d[n + 1]);
    }
}
